### ISSUE TITLE
Image Backgrounds

### ISSUE THREAD

**Issue by Cheryl-Lao** [dilevin/computer-graphics-final-image-competition#9]
I'm trying to adapt the A3 code to make a skybox, but I'm having trouble figuring out how to map images to planes. Do we need to add code to read in the image and then do a UV mapping to a box (like in A5)?

**Comment by darren-moore** [comment]
Good idea- skyboxes can add a lot to any scene. It's not built into the A3 renderer, but here's some guidance if anyone would like to implement one:

### Reading textures
I usually use `stb_image.h` for reading images. Find [here](https://github.com/nothings/stb/blob/master/stb_image.h), and cite in your readme if you use it. Drop into the `include` folder and it should be good to go.

```
#define STB_IMAGE_IMPLEMENTATION
#include "stb_image.h"

int width, height, numChannels;
unsigned char* face1 = stbi_load("cubeMapFace1.png", &width, &height, &numChannels, 0);

int r = face1[index using width, height, and numChannels]
```

### Mapping
Skyboxes can be implemented using an enormous box in the scene, but the best(and simplest) way is via [cube mapping](https://en.wikipedia.org/wiki/Cube_mapping). There's a wonderful world-vector to UV mapping function outlined on that wiki page. Now when a ray "escapes" into the void, we have an rgb value for it!

**Comment by Cheryl-Lao** [comment]
Ooo thanks Darren! I'll give it a try!

**Comment by darren-moore** [comment]
You might start with a simplified version of the mapping that just maps a single background image onto the image plane when a ray "misses". Like the A1 compositing. That might be enough for what you need.

**Comment by Cheryl-Lao** [comment]
I've been trying to debug my background compositing code for a while and I noticed that it's because the background images are being read in as just one colour. Weirdly, the value of face1[i, j, 0] seems to be the same throughout the file. The same thing is happening with JPGs and PNGs of different sizes.

I just copied your code snippet into main.cpp and changed the filename to my file. Did you run into something similar or know why that might be happening?

**Comment by darren-moore** [comment]
Some thoughts:
* When you say `face1[i, j, 0]`, make sure you're computing the correct index in the linear array. I think it's stored the same way as we write to `rgb_image` in `main.cpp`, but take care with the number of channels: transparent `.png` have 4 rgba, rather than the 3 used for `rgb_image`.
* One gotcha is that the `unsigned char` values are in the 0-255 range rather than 0-1 like the `raycolour` output. You may need to divide into the 0-1 range.
    * If you do this, `int` truncation might bite you.
* Start by simply reading and outputing a copy of a say, 100x100 png. Forgetting about any compositing for now. This is what I got to work.

**Comment by Cheryl-Lao** [comment]
Thanks for all the suggestions, I got it to work! :D It turns out I was indexing into the image array incorrectly (I was using [i, j, 0] instead of 0+3*(j+width*i)) and it was way easier to find when I just focused on copying the image
